Add support for unblocking waiting changes in wait rule conditions

praveenrewar opened the pull request:
Fixes: #427 

`unblockChanges` can be set true for a condition in wait rules of a resource, which will unblock all resources that are dependent on this resource.

`Example`
```
apiVersion: kapp.k14s.io/v1alpha1
kind: Config
waitRules:
- supportsObservedGeneration: true
  conditionMatchers:
  - type: Progressing
    status: "True"
    success: true
    unblockChanges: true
  - type: Available
    status: "True"
    success: true
  resourceMatchers:
  - apiVersionKindMatcher: {apiVersion: apps/v1, kind: Deployment}
---
apiVersion: v1
kind: Service
metadata:
  name: simple-app
  annotations:
    kapp.k14s.io/change-rule: "upsert after upserting dep"
...
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: simple-app
  annotations:
    kapp.k14s.io/change-group: "dep"
...
```

In the above example, the service will be created after the deployment, based on the change group and change rules. But the `Progressing` condition on the deployment has the `unblockChanges` set to true (in the wait rules provided as part of config) and therefore the service will be created as soon as that condition is met.

praveenrewar committed:
Add support for unblocking changes in wait rules

praveenrewar commented on pkg/kapp/config/config.go:
```diff
@@ -52,6 +52,7 @@ type WaitRuleConditionMatcher struct {
 	Failure                    bool
 	Success                    bool
 	SupportsObservedGeneration bool
+	SupportsUnblockingChanges  bool
```
Open to suggestions on choosing a better name for this. (Other alternative I am inclined towards is `unblockBlockedChanges`)

100mik replied:
Might be missing context here, when did we move away from "non-blocking"

praveenrewar replied:
I closed that PR sometime back as it wasn't aligning with the ask on issue. [This](https://github.com/vmware-tanzu/carvel-kapp/issues/427#issuecomment-1092355922) comment might help in getting to know more about the need.

sethiyash reviewed (approved):
LGTM!

cppforlife commented on pkg/kapp/resourcesmisc/custom_waiting_resource.go:
```diff
@@ -96,6 +98,12 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
 					continue
 				}
 				if condMatcher.Success {
+					if condMatcher.SupportsUnblockingChanges {
```
location of this impl seems to imply that it only works with custom waiting resource [0] which potentially becomes problematic if you want to "unblock" x but still want regular waiting logic. any thoughts?

[0] https://github.com/vmware-tanzu/carvel-kapp/blob/develop/pkg/kapp/clusterapply/converged_resource_factory.go#L47

praveenrewar replied:
So in the e2e test example, we should not be forced to add the `Available` condition for the deployment as that's something we already have for deployments, right?
Hmm..., adding that condition did bother me but I didn't worry much about it thinking that it would mostly be used for custom resources. But this is definitely a valid concern, I will think more about it.

cppforlife replied:
> So in the e2e test example, we should not be forced to add the Available condition for the deployment as that's something we already have for deployments, right?

as soon as you use custom waiting, code wont use "default" waiting rules (like for Deployment)

praveenrewar replied:
Yep, and I am not able to think of a simple way to make this work for all resources (not just custom waiting resources) using `waitRules` because of this. 
Maybe we want to introduce some sort of `unblockingRules` that are only used for unblocking blocked changes when one of the mentioned conditions matches? (Although this sounds a bit redundant from user perspective, so maybe not the best way to go about it)

praveenrewar replied:
We discussed to go forward with this approach for now and work on updating the logic for custom waiting resources so that they can be used along with default waiting rules available for other resources. I will create an issue for this and resolve this thread.

praveenrewar replied:
Created a new [issue](https://github.com/vmware-tanzu/carvel-kapp/issues/512) to track this enhancement.

cppforlife commented on pkg/kapp/resourcesmisc/custom_waiting_resource.go:
```diff
@@ -96,6 +98,12 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
 					continue
 				}
 				if condMatcher.Success {
+					if condMatcher.SupportsUnblockingChanges {
```
seems like adding this here is not enough if we would want to support same feature from ytt wait rules as well. (seems a bit awkward if we do not).

praveenrewar replied:
Hmm, I didn't consider the ytt wait rules. I will make some changes.

praveenrewar replied:
Done.

cppforlife commented on pkg/kapp/resourcesmisc/done_apply_state.go:
```diff
@@ -4,9 +4,11 @@
-	Done       bool   `json:"done"`
-	Successful bool   `json:"successful"`
-	Message    string `json:"message"`
+	Done       bool
```
json marks should be retained as they were added for ytt waiting.

praveenrewar replied:
I guess I messed up while rebasing. Good catch!!

sethiyash reviewed (commented):
Changes looks good to me. 
Should be also start using `unblockBlockedChanges` field in ytt waitRules e2e testcases (custom_wait_rules_test.go)

praveenrewar commented:
> Should be also start using unblockBlockedChanges field in ytt waitRules e2e testcases (custom_wait_rules_test.go)

Yeah, I am still trying to figure out the best way to have an e2e test for this with ytt wait rules.

sethiyash reviewed (approved):
Looks Good.

cppforlife commented on pkg/kapp/resourcesmisc/custom_waiting_resource.go:
```diff
@@ -96,6 +98,12 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
 					continue
 				}
 				if condMatcher.Success {
+					if condMatcher.SupportsUnblockingChanges {
```
hmm, i just reread this section and i'm actually stuck now on why we are "unblocking" from the success block. if it's successful then it would be considered as done (and if it's done, it's unblocked). do we need to have non-success/non-failure condition that only allows to unblock?

cppforlife replied:
(e.g. in ytt, i could say im not "done", but i will unblock anyway)

praveenrewar replied:
This was a great callout. Thank you so much!!
I have made the changes accordingly.

cppforlife commented on pkg/kapp/resourcesmisc/custom_waiting_resource.go:
```diff
@@ -95,6 +97,14 @@ func (s CustomWaitingResource) IsDoneApplying() DoneApplyState {
 					continue
 				}
+
+				if condMatcher.SupportsUnblockingChanges {
```
we should probably do this section after the below if success section. if it's successful, we just want to get out as quick as possible.

praveenrewar replied:
Totally makes sense.

praveenrewar replied:
Done

cppforlife commented on test/e2e/custom_wait_rules_test.go:
```diff
@@ -119,3 +120,144 @@ status:
+<replaced>: ---- applying 1 changes [1/3 done] ----
+<replaced>: create crontab/my-new-cron-object-1 (stable.example.com/v1) namespace: kapp-test
+<replaced>: ---- waiting on 1 changes [1/3 done] ----
+<replaced>:  ^ Unblock blocked changes
```
hmm, may be we should have some consistency between ytt and non-ytt unblock messaging. e.g.

`Allowing blocked changes to proceed: <ytt user message>`
`Allowing blocked changes to proceed: Encountered condition %s == %s: reason`

there should also be a line describing resource above the `^`. but there isnt in this output.

praveenrewar replied:
> hmm, may be we should have some consistency between ytt and non-ytt unblock messaging. e.g.

Good idea. Should we do this for other messages as well (success and failure)?

> there should also be a line describing resource above the ^. but there isnt in this output.

Apparently, the order of lines is messed up (and it's consistent, so I am guessing it's happening in one of the replace functions). I will take a look at it.

praveenrewar replied:
> hmm, may be we should have some consistency between ytt and non-ytt unblock messaging. e.g.

Done.

praveenrewar replied:
> Apparently, the order of lines is messed up (and it's consistent, so I am guessing it's happening in one of the replace functions).

`sortActionsInResult` was causing this, not sure why I added it in the first place (maybe I just copy pasted it from another test). I have removed it and the order if fixed now.

praveenrewar committed:
Add unblockBlockedChanges support to ytt waitRules

praveenrewar committed:
Support UnblockingChanges for non success/fail conditions

Also use unblockChanges instead supportUnblockingChanges and unblockBlockedChanges

cppforlife commented on test/e2e/order_test.go:
```diff
@@ -210,6 +210,102 @@ Succeeded
+    supportsUnblockingChanges: true
+  - type: Available
+    status: "True"
+    success: true
```
since this field is called success/failure=true, should we change the name of the field from `supportsUnblockingChanges` to `unblockChanges` (supports seems like a wrong naming since it's not about supporting anything).

praveenrewar replied:
Yes, makes sense. I will make the change.

praveenrewar replied:
Done.
